src/mesh_tools.rs

```diff
@@ -0,0 +1,196 @@
+use std::collections::{HashMap, HashSet};
+
+use smallvec::SmallVec;
+
+use crate::convert::cast_usize;
+use crate::geometry::Geometry;
+use crate::mesh_topology_analysis::face_to_face_topology;
+
+/// Crawls the geometry to find continuous patches of geometry.
+/// Returns a vector of new separated geometries.
+#[allow(dead_code)]
+pub fn separate_isolated_meshes(geometry: &Geometry) -> Vec<Geometry> {
+    let face_to_face = face_to_face_topology(geometry);
+    let mut available_face_indices: HashSet<u32> = face_to_face.keys().cloned().collect();
+    let mut patches: Vec<Geometry> = Vec::new();
+
+    while let Some(first_face_index) = available_face_indices.iter().next() {
+        let connected_indices = crawl_faces(*first_face_index, &face_to_face);
+
+        patches.push(
+            Geometry::from_faces_with_vertices_and_normals_remove_orphans(
+                connected_indices
+                    .iter()
+                    .map(|face_index| geometry.faces()[cast_usize(*face_index)]),
+                geometry.vertices().to_vec(),
+                geometry.normals().to_vec(),
+            ),
+        );
+
+        for c in &connected_indices {
+            available_face_indices.remove(c);
+        }
+    }
+
+    patches
+}
+
+fn crawl_faces(
+    start_face_index: u32,
+    face_to_face: &HashMap<u32, SmallVec<[u32; 8]>>,
+) -> HashSet<u32> {
+    let mut index_stack = Vec::with_capacity(face_to_face.len() / 2);
+    index_stack.push(start_face_index);
+
+    let mut connected_face_indices = HashSet::with_capacity(face_to_face.len());
+
+    while !index_stack.is_empty() {
+        let current_face_index = index_stack.pop().unwrap();
+
+        if !connected_face_indices.contains(&current_face_index) {
+            connected_face_indices.insert(current_face_index);
```

Comment:
Rust nit: `HashSet::insert` returns whether the call caused the insertion or not, so this can be simplified to save some electricity:
```rust
if connected_face_indices.insert(current_face_index) {
    // ...
}
```
https://doc.rust-lang.org/std/collections/struct.HashSet.html#method.insert